### PR TITLE
uct_ep_flush(CANCEL) non blocking

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/uct/ib/base/ib_device.c [hunk 1]
```diff
@@ -457,6 +576,13 @@ void uct_ib_device_cleanup(uct_ib_device_t *dev)
 
     ucs_debug("destroying ib device %s", uct_ib_device_name(dev));
 
+    kh_destroy_inplace(uct_ib_async_event, &dev->async_events_hash);
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:579]: leak check

File: src/uct/ib/base/ib_device.c [hunk 2]
```diff
@@ -176,6 +195,104 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+    event.resource_id = resource_id;
+    ucs_spin_lock(&dev->async_event_lock);
+    iter = kh_put(uct_ib_async_event, &dev->async_events_hash, event, &ret);
+    entry = &kh_value(&dev->async_events_hash, iter);
+    if (ret == 0) {
+        ucs_assert(entry->count > 0);
+        kh_del(uct_ib_async_event, &dev->async_events_hash, iter);
+        status       = UCS_OK;
+    } else if ((ret == 1) || (ret == 2)) {
+        entry->count = 0;
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:246]: only flag

File: src/uct/ib/base/ib_device.c [hunk 3]
```diff
@@ -176,6 +195,104 @@ static void uct_ib_device_get_locality(const char *dev_name,
...
+    }
+    ucs_spin_unlock(&dev->async_event_lock);
+}
+
+ucs_status_t
+uct_ib_device_async_event_subscribe(uct_ib_device_t *dev,
+                                    enum ibv_event_type event_type,
+                                    uint32_t resource_id,
+                                    ucs_callbackq_t *cbq,
+                                    ucs_callback_t cb, void *arg)
```
Comment (yosefe) [inline src/uct/ib/base/ib_device.c:228]: maybe use uct_completion_t instead of cb/arg

**Comment by Artemy-Mellanox** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
@Artemy-Mellanox can we close it?

**Pull request closed without merging**